app/services/bulk_hosts_manager.rb

```diff
@@ -38,4 +38,17 @@ def rebuild_configuration
     end
     all_fails
   end
+
+  def assign_taxonomy(taxonomy, optimistic_import)
+    tax_type = taxonomy.type.downcase
+    if optimistic_import == 'yes'
+      @hosts.update_all("#{tax_type}_id".to_sym => taxonomy.id)
+      # hosts location needs to be updated before import missing ids
+      taxonomy.import_missing_ids
+    elsif taxonomy.need_to_be_selected_ids.count == 0
```

Comment:
Count is fairly expensive and `!any?` should be cheaper.
```suggestion
    elsif !taxonomy.need_to_be_selected_ids.any?
```

Reply:
This refactor code is really complicated. I'd like to not touch it at all.
But this change seems simple. So here it is.

Reply:
Feel it is not straightforward to read `!my_array.any?`. Look what Google says.
![any'](https://github.com/user-attachments/assets/bbe08783-bad0-493d-b78f-a95d454eb148)

I put back the original code.

Reply:
Looks like this is a `HashWithIndifferentAccess` so it's not a query as I assumed. That means it probably doesn't matter all that much.

> Feel it is not straightforward to read `!my_array.any?`. Look what Google says.

Source? https://stackoverflow.com/questions/53480273/are-there-performance-reasons-to-prefer-size-over-length-or-count-in-ruby/53480585#53480585 says

> For Hash is not redefined and will fallback to [Enumerable#count](https://ruby-doc.org/core-2.5.3/Enumerable.html#method-i-count), which means its complexity will be O(n) as all key-values will be traversed.For Hash is not redefined and will fallback to [Enumerable#count](https://ruby-doc.org/core-2.5.3/Enumerable.html#method-i-count), which means its complexity will be O(n) as all key-values will be traversed.

Note your screenshot states:

> collections with a count property (e.g., Arrays)

But `Hash` is not such a collection if my quick research is correct.

Note that `Hash` implements [enumerable](https://rubyapi.org/o/enumerable) so you can also write:
```suggestion
    elsif taxonomy.need_to_be_selected_ids.none?
```

Overall, I think that should win on both performance and readability.